File: Ada/livro-ds/heap.c
/* (C) 1999, Arthur Vargas Lopes */
/* Arquivo fonte Heap.c          */

#include "Heap.H"

/* Algoritmo 7.5 */
void Inicializa(Heap *H)
{
	H -> Q    = 0;
	H -> V[0] = MValor;

} /* Inicializa */


/* Algoritmo 7.6 */
void Insere(Heap *H, Tipo_Dado Va)
{
	int I;

	if (H -> Q == N)  {
		printf("*** Heap overflow.\n");
		exit(1);
		}

	H -> Q         = H -> Q + 1;
	H -> V[H -> Q] = Va;
	I              = H -> Q;
	while(H -> V[I / 2] <= Va) {
		H -> V[I] = H -> V[I / 2];
		I         = I / 2;
		}
	H -> V[I] = Va;

} /* Insere */


/* Algoritmo 7.7 */
Tipo_Dado Remove(Heap *H)
{
	int J, K;
	Tipo_Dado Va, X;

	if (H -> Q == 0)  {
		printf("*** Heap underflow.\n");
		exit(1);
		}

	X         = H -> V[1];
	H -> V[1] = H -> V[H -> Q];
	H -> Q    = H -> Q - 1;
	Va        = H -> V[1];
	K         = 1;
	while(K <= (H -> Q / 2)) {
		J = K + K;
		if (J < H -> Q)
			if (H -> V[J] < H -> V[J+1])
				J = J + 1;
		if (Va >= H -> V[J])
			break;
		else {
			H -> V[K] = H -> V[J];
			K      = J;
			}
		}

	H -> V[K] = Va;
	return X;

} /* Remove */

	
/* Algoritmo 7.8 */
void Imprime(Heap *H, char Msg[])
{
	int I, J;
	int Nivel;
	char format[5];

	printf("%s\n", Msg);
	Nivel = 1;
	printf("%30s%d\n", " ", H -> V[1]);
	I     = Nivel + 1;
	while(I <= H -> Q) {
		J = 1;
		while((J <= Nivel * 2) && (I <= H -> Q))  {
			sprintf(format,"%c%ds", '%', 80 / (Nivel * 2 + 1));
			printf(format, " ");
			printf("%d", H -> V[I]);
			I = I + 1;
			J = J + 1;
			}
		putchar('\n');
		Nivel = Nivel + 1;
		}
} /* Imprime */

File: Ada/livro-ds/heap.pas
{ (C) 1999, Arthur Vargas Lopes }
{ Arquivo fonte Heap.Pas        }

PROGRAM Heap;

	USES Crt;

CONST
	N      = 12;
	MValor = 99;
TYPE
	Tipo_Dado = INTEGER;
	Heap = RECORD
		V : ARRAY[0 .. N] OF Tipo_Dado;
		Q : INTEGER;
		END;

VAR
	H1 : Heap;

	{ Algoritmo 7.5 }
	PROCEDURE Inicializa(VAR H : Heap);
	BEGIN
		H.Q    := 0;
		H.V[0] := MValor;

	END; { Inicializa }

	{ Algoritmo 7.6 }
	PROCEDURE Insere(VAR H : Heap; Va : Tipo_Dado);
	VAR
		I : INTEGER;
	BEGIN
		IF H.Q = N THEN BEGIN
			Writeln('*** Heap overflow.');
			Halt;
			END;
		H.Q      := H.Q + 1;
		H.V[H.Q] := Va;
		I        := H.Q;
		WHILE H.V[I DIV 2] <= Va DO BEGIN
			H.V[I] := H.V[I DIV 2];
			I      := I DIV 2
			END;
		H.V[I] := Va;

	END; { Insere }

	{ Algoritmo 7.7 }
	FUNCTION Remove(VAR H : Heap) : Tipo_Dado;
	VAR
		J, K    : INTEGER;
		Va      : Tipo_Dado;
		Termina : BOOLEAN;
	BEGIN
		IF H.Q = 0 THEN BEGIN
			Writeln('*** Heap underflow.');
			Halt;
			END;
		Remove  := H.V[1];
		H.V[1]  := H.V[H.Q];
		H.Q     := H.Q - 1;
		Termina := FALSE;
		Va      := H.V[1];
		K       := 1;
		WHILE (K <= (H.Q DIV 2)) AND
		      (NOT Termina) DO BEGIN
			J := K + K;
			IF J < H.Q THEN
				IF H.V[J] < H.V[J+1] THEN
					J := J + 1;
			IF Va >= H.V[J] THEN
				Termina := TRUE
			ELSE BEGIN
				H.V[K] := H.V[J];
				K      := J;
				END;
			END;
		H.V[K] := Va;
	END; { Remove }

	{ Algoritmo 7.8 }
	PROCEDURE Imprime(H : Heap; Msg : STRING);
	VAR
		I, J  : INTEGER;
		Nivel : INTEGER;

	BEGIN
		Writeln(Msg);
		Nivel := 1;
		Writeln(H.V[1] : 39);
		I     := Nivel + 1;
		WHILE I <= H.Q DO BEGIN
			J := 1;
			WHILE (J <= Nivel * 2) AND (I <= H.Q) DO BEGIN
				Write(H.V[I] : 80 DIV (Nivel * 2 + 1));
				I := I + 1;
				J := J + 1;
				END;
			Writeln;
			Nivel := Nivel + 1;
			END;
	END; { Imprime }
BEGIN
	Inicializa(H1);
	Insere(H1,40);
	Insere(H1,12);
	Insere(H1,35);
	Insere(H1,15);
	Insere(H1,2);
	Insere(H1,20);
	Insere(H1,17);
	Insere(H1,19);
	Insere(H1,8);
	Imprime(H1,'Heap apos a insercao de 40,12,35,2,20,17,8,19,15:');
	Writeln('Remove(H1) = ', Remove(H1));
	Imprime(H1,'Heap apos remocao do maior elemento:');

END. { Heap }